Add NFT proposal (#143)
aeternity/AEXs: AEXS/aex-141.md

# AEX Non-Fungible Token Standard

```
AEX: 141
Title: Non-Fungible Token Standard
Author: Arjan van Eersel <arjan@appiness.solutions> (@zkvonsnarkenstein)
License: ISC

> Discussions-To: <URL>
> Status: Draft

Reviewer: ```suggestion
Discussions-To: https://forum.aeternity.com/t/aeternity-nft-token-standard/9781
Status: Review
```

A standard implementation of non-fungible tokens for the Aeternity ecosystem. The design goal of the primary interface is to be as compatible with ERC-721 as possible, so that anyone who can work with ERC-721 can work with this interface. However, where Sophia offers a better way, performance and efficiency should prevail over compatibility.

There is no support for unsafe transactions. Therefore all transactions are ought to be safe.

## Motivation

The following standard describes standard interfaces for non-fungible tokens. The proposal contains a primary interface and secondary interfaces for optional functionality that not everyone might need. 

# Basic NFT

## Interface

```sophia
contract interface NFT =
    datatype metadata_type = URL | IPFS | OBJECT_ID | MAP
    datatype metadata = string | map(string,string)

    record meta_info = 
        { name: string
        , symbol: string 
        , base_url: option(string)
        , type : metadata_type)}

    datatype event 
        = Transfer(indexed address, indexed address, indexed int)
        | Approval(indexed address, indexed address, indexed int, bool)
        | ApprovalForAll(indexed address, indexed address, bool)

>     entrypoint aexX_extensions() : list(string)

Reviewer: ```suggestion
    entrypoint aex141_extensions() : list(string)
```

    entrypoint meta_info() : meta_info

    entrypoint metadata(int): option(metadata)

    stateful entrypoint mint : (address, option(string)) => int

    entrypoint balance : (address) => option(int)

    entrypoint owner : (int) => option(address)
        
    stateful transfer : (from address, to address, token int, data option(string)) => unit

    stateful entrypoint approve : (address, int, bool) => unit

    stateful entrypoint approve_all : (address, bool) => unit

    entrypoint get_approved : (int) => option(address)

    entrypoint is_approved : (int, address) => bool

    entrypoint is_approved_for_all : (address, address) => bool
```

## Methods

> ### aexX_extensions\(\)
> 
> This function **returns** a hardcoded list of all implemented extensions on the deployed contract.
> X is to be replaced with the approriate number allocated to this AEX proposal. At this stage the number is unknown.
> 
> ```sophia
> entrypoint aex9_extensions() : list(string)
> ```

Reviewer: ```suggestion
### aex141_extensions\(\)

This function **returns** a hardcoded list of all implemented extensions on the deployed contract.

```sophia
entrypoint aex141_extensions() : list(string)
```
```

> # AEX Non-Fungible Token Standard

Reviewer: ```suggestion
# AEX-141: Non-Fungible Token Standard
```
